### DISCUSSION TITLE
Unable to deploy dockerized FastAPI application to ECS fargate

### DISCUSSION THREAD

**Discussion by schrecka** [fastapi/fastapi#8811] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Dockerfile:
# Image from dockerhub
FROM python:3.8.1-slim 

ENV PYTHONUNBUFFERED 1 
# Expose the port 8000 in which our application runs
EXPOSE 8000 
# Make /app as a working directory in the container
WORKDIR /app 
# Copy requirements from host, to docker container in /app 
COPY ./requirements.txt .
# Copy everything from ./app directory to /app in the container
COPY ./app . 
# Install the dependencies
RUN pip install --no-deps -r requirements.txt
# Run the application in the port 8000
ENTRYPOINT [ "uvicorn" ]
CMD ["--host", "0.0.0.0", "--port", "8000", "app.main:app"]

docker-compose.yml
version: '3.8'

services:
  web:
    restart: always
    build:
      context: .
      dockerfile: ./Dockerfile
    volumes:
      - .:/app
    ports:
      - 8000:8000
```

### Description

docker-compose and docker build both work locally on Docker Desktop. However, when I deploy to ECS/fargate, I get the following error. 

**### Cloudwatch logs for ECS/Fargate Task**

Traceback (most recent call last):
--
File "/usr/local/bin/uvicorn", line 8, in <module>
sys.exit(main())
File "/usr/local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
return self.main(*args, **kwargs)
File "/usr/local/lib/python3.8/site-packages/click/core.py", line 782, in main
rv = self.invoke(ctx)
File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
return ctx.invoke(self.callback, **ctx.params)
File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
return callback(*args, **kwargs)
File "/usr/local/lib/python3.8/site-packages/uvicorn/main.py", line 362, in main
run(**kwargs)
File "/usr/local/lib/python3.8/site-packages/uvicorn/main.py", line 386, in run
server.run()
File "/usr/local/lib/python3.8/site-packages/uvicorn/server.py", line 49, in run
loop.run_until_complete(self.serve(sockets=sockets))
File "uvloop/loop.pyx", line 1494, in uvloop.loop.Loop.run_until_complete
File "/usr/local/lib/python3.8/site-packages/uvicorn/server.py", line 56, in serve
config.load()
File "/usr/local/lib/python3.8/site-packages/uvicorn/config.py", line 308, in load
self.loaded_app = import_from_string(self.app)
File "/usr/local/lib/python3.8/site-packages/uvicorn/importer.py", line 23, in import_from_string
raise exc from None
File "/usr/local/lib/python3.8/site-packages/uvicorn/importer.py", line 20, in import_from_string
module = importlib.import_module(module_str)
File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
return _bootstrap._gcd_import(name[level:], package, level)
File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
File "<frozen importlib._bootstrap>", line 991, in _find_and_load
File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
File "<frozen importlib._bootstrap>", line 991, in _find_and_load
File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'uvicorn app'

### Operating System

Linux

### Operating System Details

docker

### FastAPI Version

0.65.2

### Python Version

3.8.1

**Comment by Honda-a** [comment]
Can you share your requirements.txt file ?

**Comment by schrecka** [comment]
> Can you share your requirements.txt file ?

aiofiles==0.5.0
aniso8601==7.0.0
async-exit-stack==1.0.1
async-generator==1.10
beautifulsoup4==4.9.3
certifi==2021.5.30
chardet==4.0.0
click==7.1.2
dateparser==0.7.6
dnspython==2.1.0
email-validator==1.1.3
fastapi==0.65.2
feedparser==6.0.8
graphene==2.1.8
graphql-core==2.3.2
graphql-relay==2.0.1
h11==0.12.0
httptools==0.1.2
idna==2.10
itsdangerous==1.1.0
Jinja2==2.11.3
MarkupSafe==2.0.1
orjson==3.5.3
promise==2.3
pydantic==1.8.2
pygooglenews==0.1.2
python-dateutil==2.8.1
python-dotenv==0.17.1
python-multipart==0.0.5
pytz==2021.1
PyYAML==5.4.1
regex==2021.4.4
requests==2.25.1
Rx==1.6.1
sgmllib3k==1.0.0
six==1.16.0
soupsieve==2.2.1
starlette==0.14.2
typing-extensions==3.10.0.0
tzlocal==2.1
ujson==4.0.2
urllib3==1.26.5
uvicorn==0.13.4
uvloop==0.15.2
watchgod==0.7
websockets==8.1

**Comment by ilyaGotfryd** [comment]
Might be a naive suggestion but does your python path for the `app` end up being correct? I feel like your docker-compose ends up with the `/app/app` path and has content in that second app folder. At the same time your docker build just has an `/app` folder with content in it and you are launching the app with `app.main` python path from `WORKDIR /app` that would not resolve.

**Comment by rahulbprakash-mv** [comment]
This seems to be an issue with the python path not being set.
https://github.com/tiangolo/fastapi/issues/2582

**Comment by schrecka** [comment]
> This seems to be an issue with the python path not being set.
> #2582

I added `ENV PYTHONPATH "${PYTHONPATH}:\$PWD"` to the beginning of my Dockerfile. Still the same error

**Comment by schrecka** [comment]
> Might be a naive suggestion but does your python path for the `app` end up being correct? I feel like your docker-compose ends up with the `/app/app` path and has content in that second app folder. At the same time your docker build just has an `/app` folder with content in it and you are launching the app with `app.main` python path from `WORKDIR /app` that would not resolve.

Hm ok I can try changing the path. Only thing is how would my service build on DockerDesktop if I had an error like this?

**Comment by Honda-a** [comment]
>Hm ok I can try changing the path. Only thing is how would my service build on DockerDesktop if I had an error like this?

fargate is kind of like magic things dont work on it for some reason i had the same problem before as well with other Applications. It would work perfectly on local but on fargate it would give error

also there seems to be a problem with builds on MacOs but you are linux so it has nothing to do with you

**Comment by iamthen0ise** [comment]
Hi @schrecka 
I am confused by the error text itself:

```
 ModuleNotFoundError: No module named 'uvicorn app'
```

Is it a good practice to use entrypoint command like this? Usually I do not use `ENTRYPOINT`for binary files.
```
ENTRYPOINT [ "uvicorn" ]
CMD ["--host", "0.0.0.0", "--port", "8000", "app.main:app"]
```

Would it work if you will run application as it mentioned in [FastAPI docs](https://fastapi.tiangolo.com/deployment/docker/#raspberry-pi-and-other-architectures)? The example provided for RaspberryPI, but mentioned `CMD` command is valid for any platform.
```
CMD ["uvicorn", "app.main:app",  "--host", "0.0.0.0", "--port", "8000"]
```

Sorry if i missed something

**Comment by zavbala** [comment]
Hello @schrecka, were you able to deploy it? I have many issues too
